### ISSUE TITLE
Support entropy from RDSEED

### ISSUE THREAD

**Issue by djs55** [mirage/mirage-entropy#9]
On modern Intel boxes `RDSEED` is a good source of entropy (probably)

**Comment by dsheets** [comment]
See also [TimKnast/ocaml-rdseed](https://github.com/TimKnast/ocaml-rdseed).

**Comment by pqwy** [comment]
It is good, and even `RDRAND` is (it runs `AES-CTR` for no more than 511 consecutive steps reseeding it with what became `RDSEED`). And I have two papers arguing that at least on x86, the CPU is so inherently prone to data races due to several levels of paralelism and caching effects, that its physical behaviour can reasonably be treated as nondeterministic -- which suggests reading the cycle counter every now and then should yield a steady trickle of entropy. These two things are used by both [haveged](http://www.issihosts.com/haveged/) and [Whirlwind RNG](http://www.ieee-security.org/TC/SP2014/papers/Not-So-RandomNumbersinVirtualizedLinuxandtheWhirlwindRNG.pdf).

What's even better, both instructions can be exposed to OCaml with about 4 lines of C each (returning `Val_long` and bypassing GC just to be super-awesome).

What's not so good is that both these sources (*especially* the cycle counter) need to be checked periodically, as a single sample is not sufficient (and in general the rng is built to recover from state compromise *provided* there is a steady influx of entropy). This probably calls for piggy-backing on the event mechanism already in place, to avoid constantly waking up the system in the background.

The [interface](http://ocsigen.org/lwt/2.4.7/api/Lwt_engine) is too constrained for that, but it can be [worked around](https://github.com/mirleft/ocaml-tls/blob/master/lwt/tls_lwt.ml#L288) with a fragile hack. With a hook in place, of type `unit -> unit` that gets called whenever `Lwt` is ready to dispatch an event, adding these two sources of entropy is a matter of 15 minutes.

The real problem here is what would a more general interface for trapping any activity in an `Lwt_engine.t` look like? Pre-fire? Post-fire? Controllable? With a single callback? A queue? Etc. We need to come up with a coherent design before PR-ing that to lwt people.

And the really ugly part is the fact that various platforms can expect wildly varying support for different mechanisms of obtaining entropy. How do we expose that to `config.ml`?

Having `one_entropy` and `another_entropy`, where the blue one reads from `xentropyd` and the red one uses builtin instructions? What happens with processors that don't support some of the instructions? Disable the red one at compile-time? What does compositionality of those sources look like? How can they be mixed?

**TL;DR**
What's stopping us from having proper entropy is the lack of clarity on how to describe source combinations in `config.ml`, how to patch `Lwt_engine` and how the two should interact.

**Comment by mor1** [comment]
i can't speak to how best to patch `Lwt_engine` but regarding describing
source combinations in `config.ml`: perhaps one should view entropy as
sufficiently important that it must be explicitly specified by the
developer and that's that?

that is, if she wants unikernels that work on different platforms, she
compiles different images (`--entropy-aws`, `--entropy-xentropyd`, etc if i
have the right idea of @nojb's patches to mirage cli). if she wants
complicated behaviour, she can code it up as she likes-- but we don't try
to second guess *anything*.


On 2 February 2015 at 21:08, David Kaloper <notifications@github.com> wrote:

> It is good, and even RDRAND is (it runs AES-CTR for no more than 511
> consecutive steps reseeding it with what became RDSEED). And I have two
> papers arguing that at least on x86, the CPU is so inherently prone to data
> races due to several levels of paralelism and caching effects, that its
> physical behaviour can reasonably be treated as nondeterministic -- which
> suggests reading the cycle counter every now and then should yield a steady
> trickle of entropy. These two things are used by both haveged
> <http://www.issihosts.com/haveged/> and Whirlwind RNG
> <http://www.ieee-security.org/TC/SP2014/papers/Not-So-RandomNumbersinVirtualizedLinuxandtheWhirlwindRNG.pdf>
> .
>
> What's even better, both instructions can be exposed to OCaml with about 4
> lines of C each (returning Val_long and bypassing GC just to be
> super-awesome).
>
> What's not so good is that both these sources (*especially* the cycle
> counter) need to be checked periodically, as a single sample is not
> sufficient (and in general the rng is built to recover from state
> compromise *provided* there is a steady influx of entropy). This probably
> calls for piggy-backing on the event mechanism already in place, to avoid
> constantly waking up the system in the background.
>
> The interface <http://ocsigen.org/lwt/2.4.7/api/Lwt_engine> is too
> constrained for that, but it can be worked around
> <https://github.com/mirleft/ocaml-tls/blob/master/lwt/tls_lwt.ml#L288>
> with a fragile hack. With a hook in place, of type unit -> unit that gets
> called whenever Lwt is ready to dispatch an event, adding these two
> sources of entropy is a matter of 15 minutes.
>
> The real problem here is what would a more general interface for trapping
> any activity in an Lwt_engine.t look like? Pre-fire? Post-fire?
> Controllable? With a single callback? A queue? Etc. We need to come up with
> a coherent design before PR-ing that to lwt people.
>
> And the really ugly part is the fact that various platforms can expect
> wildly varying support for different mechanisms of obtaining entropy. How
> do we expose that to config.ml?
>
> Having one_entropy and another_entropy, where the blue one reads from
> xentropyd and the red one uses builtin instructions? What happens with
> processors that don't support some of the instructions? Disable the red one
> at compile-time? What does compositionality of those sources look like? How
> can they be mixed?
>
> *TL;DR*
> What's stopping us from having proper entropy is the lack of clarity on
> how to describe source combinations in config.ml, how to patch Lwt_engine
> and how the two should interact.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-entropy/issues/9#issuecomment-72538013>.
>



-- 
Richard Mortier
mort@cantab.net

**Comment by pqwy** [comment]
That's the direction I would take too, of being completely explicit.

But then the question of composition becomes even more important. There must be a design for expressing things like `RDENT`, `TIMER+Xentropyd` or `TIMER+RDENT+Xentropyd`.

**Comment by mor1** [comment]
let a million such designs bloom; if some turn out to be great, the Magic
of FOSS will see to it they survive. (right? :))

but i don't think we should try to specify up front, and (assuming the
atoms are of use) we shouldn't hold things up while we try and work it out
ourselves...

On 2 February 2015 at 21:33, David Kaloper <notifications@github.com> wrote:

> That's the direction I would take too, of being completely explicit.
>
> But then the question of composition becomes even more important. There
> must be a design for expressing things like RDENT, TIMER+Xentropyd or
> TIMER+RDENT+Xentropyd.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-entropy/issues/9#issuecomment-72542471>.
>



-- 
Richard Mortier
mort@cantab.net

**Comment by pqwy** [comment]
Preliminary (bare) instruction [support](https://github.com/pqwy/mirage-entropy/commit/5a1a0becb2d48e07cab895d0ea257b6884b13cf5); still needs wiring into everything else.

**Comment by hannesm** [comment]
merged into master! \o/
